e2e/app/page/notebook-page.ts

```diff
@@ -395,6 +397,42 @@ export default class NotebookPage extends NotebookFrame {
     return codeOutput;
   }
 
+  async runAllCells(): Promise<void> {
+    const clickCellMenuIcon = async (iframe: Frame): Promise<void> => {
+      await iframe.waitForXPath(Xpath.cellMenuDropdown, { visible: true, timeout: 2000 }).then((element) => {
+        element.hover();
+        element.click();
+      });
+    };
+
+    let maxRetries = 3;
+    const clickAndCheck = async (iframe: Frame) => {
+      await clickCellMenuIcon(iframe);
+      const succeeded = await iframe
+        .waitForXPath(Xpath.runAllCode, { visible: true, timeout: 2000 })
+        .then((menuitem) => {
+          menuitem.hover();
+          menuitem.click();
+          return true;
+        })
+        .catch(() => {
+          return false;
+        });
+      if (succeeded) {
+        return;
+      }
+      if (maxRetries <= 0) {
+        throw new Error('Failed to click Cell menu -> Run All.');
+      }
+      maxRetries--;
+      await this.page.waitForTimeout(1000).then(() => clickAndCheck(iframe)); // 1 second pause and retry.
```

Comment:
nit: Generally I would avoid recursive code unless it is simplifying an otherwise complex abstraction. In the case of this retry loop, I think just using a regular loop would be more appropriate, e.g. 

```
for (let retries = 3; retries > 0; retries--) {
  ...
  if (succeeded) {
    return;
  }
  await this.page.waitForTimeout(1000);
}
throw new Error('Failed to click Cell menu -> Run All.');
```